### PR TITLE
Bind comment updates

### PR DESCRIPTION
**Pull request by davidge20**
## Description

Commented out bind_syscall and all functions related to it

### CODE REVIEW

File: src/safeposix/syscalls/net_calls.rs [hunk 1]
```diff
@@ -21,6 +22,10 @@ impl Cage {
         cloexec: bool,
         conn: ConnState,
     ) -> SocketDesc {
+        //For blocking sockets, operation wait until completed.
+        //For non-blocking sockets, operations return immediately, even if the requested operation 
+        //is not completed.
+        //cloexec flag closes the fd pointing to the socket on the execution of a new program
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: Why do we need to track these flags?

File: src/safeposix/syscalls/net_calls.rs [hunk 2]
```diff
@@ -36,29 +41,45 @@ impl Cage {
         return sockfd;
     }
 
+    //Find an available file descriptor index in the File Descriptor Table
+    //If available, insert the socket fd into the File Descriptor Table at the 
+    //available index
     fn _socket_inserter(&self, sockfd: FileDescriptor) -> i32 {
         let (fd, guardopt) = self.get_next_fd(None);
+        //Indicate an error in the fd and propogate the error forward
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: When would this occur?  Why?

File: src/safeposix/syscalls/net_calls.rs [hunk 3]
```diff
@@ -36,29 +41,45 @@ impl Cage {
...
+    //available index
     fn _socket_inserter(&self, sockfd: FileDescriptor) -> i32 {
         let (fd, guardopt) = self.get_next_fd(None);
+        //Indicate an error in the fd and propogate the error forward 
         if fd < 0 {
             return fd;
         }
         let fdoption = &mut *guardopt.unwrap();
-        let _insertval = fdoption.insert(sockfd);
+        //Insert the sockfd as the FileDescriptor inside fdoption
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: Why?  What are we doing here?  

(We can assume the reader knows Rust well enough to know they are inserting  a sockfd inside the fdoption datastructure)

File: src/safeposix/syscalls/net_calls.rs [hunk 4]
```diff
@@ -36,29 +41,45 @@ impl Cage {
...
     fn _implicit_bind(&self, sockhandle: &mut SocketHandle, domain: i32) -> i32 {
         if sockhandle.localaddr.is_none() {
+            //Assign a new local address to the socket handle
             let localaddr = match Self::assign_new_addr(
                 sockhandle,
                 domain,
+                //The SO_RESUEPORT bit within the protocol encodes rebind ability
+                //The rebind ability of a socket refers to whether a socket can be 
+                //re-bound to an address and port that it was previously bound to, 
+                //especially after it has been closed or if the binding has been reset.
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs:76]: Okay, why do we set this?  

It's slightly odd that SO_REUSEPORT is used to leftshift a bit.  Why is this done?

File: src/safeposix/syscalls/net_calls.rs [hunk 5]
```diff
@@ -159,35 +180,49 @@ impl Cage {
 
     //creates a sockhandle if none exists, otherwise this is a no-op
     pub fn force_innersocket(sockhandle: &mut SocketHandle) {
+        //If innersocket is not available, then insert a socket into innersocket
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: What is an innersocket?

File: src/safeposix/syscalls/net_calls.rs [hunk 6]
```diff
@@ -159,35 +180,49 @@ impl Cage {
...
         if let None = sockhandle.innersocket {
+            //Create a new socket
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: I can see we're creating a socket, but why?

Also, can this operation fail?   What happens if it does?

File: src/safeposix/syscalls/net_calls.rs [hunk 7]
```diff
@@ -159,35 +180,49 @@ impl Cage {
...
             let thissock =
                 interface::Socket::new(sockhandle.domain, sockhandle.socktype, sockhandle.protocol);
 
+            //Loop through socket options and check which ones are set
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs:201]: This is a weird little bit of code.  I think you want to explain that because you need to make separate calls to set different socket options (i.e. you can't set multiple flags at once) you loop here.

File: src/safeposix/syscalls/net_calls.rs [hunk 8]
```diff
@@ -196,6 +231,7 @@ impl Cage {
             );
         }
 
+        //If the socket is already binded to an address, exit with error
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]:
```suggestion
        //If the socket is already bound to an address, exit with error
```

File: src/safeposix/syscalls/net_calls.rs [hunk 9]
```diff
@@ -206,6 +242,7 @@ impl Cage {
 
         let mut newsockaddr = localaddr.clone();
 
+        //Bind socket based on domain type
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs:320]: Why do you need to do this check?

File: src/safeposix/syscalls/net_calls.rs [hunk 10]
```diff
@@ -244,18 +285,25 @@ impl Cage {
                 let filename = truepath.file_name().unwrap().to_str().unwrap().to_string(); //for now we assume this is sane, but maybe this should be checked later
 
                 //this may end up skipping an inode number in the case of ENOTDIR, but that's not catastrophic
+                //FS_METADATA contains information about the file system
                 let newinodenum = FS_METADATA
                     .nextinode
-                    .fetch_add(1, interface::RustAtomicOrdering::Relaxed); //fetch_add returns the previous value, which is the inode number we want
+                    .fetch_add(1, interface::RustAtomicOrdering::Relaxed);
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs:369]: Interesting.  What does Relaxed do here?

File: src/safeposix/syscalls/net_calls.rs [hunk 11]
```diff
@@ -268,26 +316,35 @@ impl Cage {
                         mtime: time,
                     });
 
+                    //Insert new inode number of file to the same file system
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: I can't understand what you mean here.  Can you explain in more detail?

File: src/safeposix/syscalls/net_calls.rs [hunk 12]
```diff
@@ -268,26 +316,35 @@ impl Cage {
...
                     dir.linkcount += 1;
                 } else {
+                    //Parent dictory inode does not exist in inode table of file system
                     return syscall_error(
                         Errno::ENOTDIR,
                         "bind",
                         "unix domain socket path made socket address child of non-directory file",
                     );
                 }
+                //Insert unix info into socket handle
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs:413]: Why?  What is this doing?

File: src/safeposix/syscalls/net_calls.rs [hunk 13]
```diff
@@ -268,26 +316,35 @@ impl Cage {
...
                     inode: newinodenum,
                 });
 
+                //Insert path to socket file into a set
                 NET_METADATA.domsock_paths.insert(truepath);
+                //Insert the file inode num and inode as key-value pair into
+                //file system inode table
                 FS_METADATA.inodetable.insert(newinodenum, newinode);
             }
+            //File already exists
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: What is this doing?!?

File: src/safeposix/syscalls/net_calls.rs [hunk 14]
```diff
@@ -296,35 +353,47 @@ impl Cage {
...
         } else {
+            //Reserve a new local port num
             let localout = NET_METADATA._reserve_localport(
                 newsockaddr.addr(),
                 newsockaddr.port(),
                 sockhandle.protocol,
                 sockhandle.domain,
                 intent_to_rebind,
             );
+            //If reserving a local port causes an error, return the error num
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: You could possibly do the next few lines as a match statement instead.  It would make it clearer that your last unwrap is safe...

I'll note, it is slightly weird that the code here really just wants the inner result regardless of if it is an err or not...

File: src/safeposix/syscalls/net_calls.rs [hunk 15]
```diff
@@ -337,27 +406,40 @@ impl Cage {
...
+    //localaddr: reference to the GenSockaddr enum that hold the address
+    //prereserved: bool that describes whether the address and port have 
+    //             been set aside or designated for specific purposes
     pub fn bind_inner(
         &self,
         fd: i32,
         localaddr: &interface::GenSockaddr,
         prereserved: bool,
     ) -> i32 {
+        //chekedfd is an atomic reference count of a lock on the fd
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]:
```suggestion
        //checkedfd is an atomic reference count of a lock on the fd
```

Of a lock?  Or of the number of locks?  Or what?

File: src/safeposix/syscalls/net_calls.rs [hunk 16]
```diff
@@ -337,27 +406,40 @@ impl Cage {
...
         let checkedfd = self.get_filedescriptor(fd).unwrap();
-        let mut unlocked_fd = checkedfd.write();
+        //returns a write lock once no other writers or readers have access to the lock
+        let mut unlocked_fd = checkedfd.write(); //why isn't .unwrap() used here?
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: Great idea to add a comment saying what you don't know!!!  😄

File: src/safeposix/syscalls/net_calls.rs [hunk 17]
```diff
@@ -337,27 +406,40 @@ impl Cage {
...
         if let Some(filedesc_enum) = &mut *unlocked_fd {
             match filedesc_enum {
+                //*unlocked_fd is in the format of Option<T>, where T is of type Socket(&mut SocketDesc)
                 Socket(ref mut sockfdobj) => {
+                    //Clone the socket handle
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs:515]: Why do we clone this?

File: src/safeposix/syscalls/net_calls.rs [hunk 18]
```diff
@@ -337,27 +406,40 @@ impl Cage {
...
                     let sock_tmp = sockfdobj.handle.clone();
+                    //Access write gaurd for socket handle
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]:
```suggestion
                    //Access write guard for socket handle
```

Does this access the write guard?  How?

**Review by JustinCappos — CHANGES_REQUESTED** [review]
Overall, it's a good first pass.  At a high level:

1) focus on why the code does something instead of what.  For example, why did you insert an item into a data structure?

2) Feel free to say when you are not sure.  This is totally normal and something I do when writing my own code.  Don't be shy to do this more often.

3) if you want to cite a source go ahead.  For example, if you read a manpage to understand something or looked at an article, it's okay to drop a link, especially if the concept is complicated.  This might be helpful for some of the setsockopt activities and related flags.

File: src/safeposix/syscalls/net_calls.rs (hunk 4 above)
Comment (davidge20) [reply src/safeposix/syscalls/net_calls.rs:76]: I found a source that explains the importance of  the SO_RESUSEPORT socket option.

https://lwn.net/Articles/542629/

I was wondering if some post like this is valid to cite since I'm not exactly sure of the credentials of the site but it seems legit

Comment (JustinCappos) [reply src/safeposix/syscalls/net_calls.rs:76]: Sure.  You should cite whatever you looked at.

**Comment by JustinCappos** [comment]
Please ping me for a review after you've had a chance to address all the comments.

**Comment by davidge20** [comment]
> Please ping me for a review after you've had a chance to address all the comments.

File: src/safeposix/syscalls/net_calls.rs [hunk 19]
```diff
@@ -12,6 +12,7 @@ use crate::safeposix::filesystem::*;
 use crate::safeposix::net::*;
 
 impl Cage {
+    //Initializes a socket description and sets the necessary flags
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs]: should be "socket file descriptor"

File: src/safeposix/syscalls/net_calls.rs [hunk 20]
```diff
@@ -36,29 +44,49 @@ impl Cage {
...
     fn _socket_inserter(&self, sockfd: FileDescriptor) -> i32 {
         let (fd, guardopt) = self.get_next_fd(None);
+        //In the case that no fd is available from the call to get_next_fd,
+        //fd is set to -ENFILE = -23 and the error is propagated forward
         if fd < 0 {
             return fd;
         }
         let fdoption = &mut *guardopt.unwrap();
-        let _insertval = fdoption.insert(sockfd);
+        //Insert the sockfd as the FileDescriptor inside fdoption
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs]: The option doesn't say anything about whats inside it, it just is None or Some(). By inserting into it indicates that there is now a fd at that index

File: src/safeposix/syscalls/net_calls.rs [hunk 21]
```diff
@@ -159,35 +187,64 @@ impl Cage {
...
+        //The innersocket is an Option wrapped around the raw sys fd
+        //Depending on domain, innersocket may or may not be necessary
+        //Ex: Unix sockets do not have a kernal fd
+        //If innersocket is not available, then create a socket and insert its
+        //fd into innersocket
         if let None = sockhandle.innersocket {
+            //Create a new socket, as no sockhandle exists
+            //Socket creation rarely fails except for invalid parameters or extremely low-resources conditions
+            //Upon failure, process will panic
+            // ** What domains does lind satisfy???? ** //
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs]: we handle ipv4, ipv6, and unix

File: src/safeposix/syscalls/net_calls.rs [hunk 22]
```diff
@@ -244,18 +310,31 @@ impl Cage {
                 let filename = truepath.file_name().unwrap().to_str().unwrap().to_string(); //for now we assume this is sane, but maybe this should be checked later
 
                 //this may end up skipping an inode number in the case of ENOTDIR, but that's not catastrophic
+                //FS_METADATA contains information about the file system
                 let newinodenum = FS_METADATA
                     .nextinode
-                    .fetch_add(1, interface::RustAtomicOrdering::Relaxed); //fetch_add returns the previous value, which is the inode number we want
+                    .fetch_add(1, interface::RustAtomicOrdering::Relaxed); 
+                    //fetch_add returns the previous value, which is the inode number we want,
```
Comment (rennergade) [inline src/safeposix/syscalls/net_calls.rs]: this is good for your understanding but im not sure this needs a lenghty comment, but ill defer to Justin

**Review by rennergade — CHANGES_REQUESTED** [review]
Requested some minor changes. Also should add some tests, bind is used in most of the net tests in the test suite already, but worth adding some explicit edge cases.

**Comment by davidge20** [comment]
@JustinCappos pinging for further review after updates

File: src/safeposix/syscalls/net_calls.rs [hunk 23]
```diff
@@ -159,35 +187,64 @@ impl Cage {
...
+
+                //Otherwise, set the socket option in the new socket
+                //The level argument specifies the protocol level at which the option resides
+                //In our case, we are setting options at the socket level
+                //To learn more about setsockopt https://man7.org/linux/man-pages/man3/setsockopt.3p.html
                 let sockret = thissock.setsockopt(SOL_SOCKET, reuse, 1);
+                //Failure occured upon setting a socket option
+                //Possible failures can be read at the man page linked above
+                //
+                // **Possibly add errors instead of having a single panick for all errors ??? ** //
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]:
```suggestion
                // TODO: Possibly add errors instead of having a single panic for all errors ??? ** //
```

File: src/safeposix/syscalls/net_calls.rs [hunk 24]
```diff
@@ -159,35 +187,64 @@ impl Cage {
...
 
     //we assume we've converted into a RustSockAddr in the dispatcher
+    //bind a name to a socket - assign the address specified by localaddr to the 
+    //socket referred to by the file descriptor fd
     pub fn bind_syscall(&self, fd: i32, localaddr: &interface::GenSockaddr) -> i32 {
         self.bind_inner(fd, localaddr, false)
     }
 
+    //Based on the domain of the socket, bind the socket with a local address and 
+    //insert a cloned local address in the socket handle
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: But what is an "inner" socket?  What is the purpose of this?

File: src/safeposix/syscalls/net_calls.rs [hunk 25]
```diff
@@ -159,35 +187,64 @@ impl Cage {
...
             }
 
+            //Insert the socket file descriptor into the innersocket value
             sockhandle.innersocket = Some(thissock);
         };
     }
 
     //we assume we've converted into a RustSockAddr in the dispatcher
+    //bind a name to a socket - assign the address specified by localaddr to the 
+    //socket referred to by the file descriptor fd
```
Comment (JustinCappos) [inline src/safeposix/syscalls/net_calls.rs]: The bind_inner function seems pointless.  Why have it?

File: src/safeposix/syscalls/net_calls.rs (hunk 22 above)
Comment (JustinCappos) [reply src/safeposix/syscalls/net_calls.rs]: I'd say the comment can be shortened, but something is needed here.  I'm actually more interested in *why* relaxed ordering was used instead of what relaxed ordering does...

**Review by JustinCappos — APPROVED** [review]
It's close.  Please fix the things I mentioned (if something isn't clear, let me know) and then have Yash or Nick merge...

File: src/safeposix/syscalls/net_calls.rs (hunk 25 above)
Comment (rennergade) [reply src/safeposix/syscalls/net_calls.rs]: I'm pretty sure this is a remnant from a refactor at some point.

File: src/safeposix/syscalls/net_calls.rs (hunk 22 above)
Comment (rennergade) [reply src/safeposix/syscalls/net_calls.rs]: We just went with the fastest ordering guaranteed to be atomic since we probably don't need to be more restrictive for a counter.

https://stackoverflow.com/questions/30407121/which-stdsyncatomicordering-to-use

**Review by rennergade — APPROVED** [review]
I think this is good to go. Approved!

**Review by JustinCappos — APPROVED** [review]
LGTM!
